### PR TITLE
@PrimaryKey + @Required annotation generates inappropriate type cast

### PR DESCRIPTION
**Pull request by stk1m1**
In issue #2637 where `@PrimaryKey` and `@Required` annotations are applied together on `String` type field, inappropriate type cast is generated to cause compilation fail. This PR addresses the issue.

@realm/java please review.

### CODE REVIEW

**Comment by stk1m1** [comment]
retest this please

**Comment by stk1m1** [comment]
retest this please

**Comment by stk1m1** [comment]
retest this please

File: realm/realm-library/src/androidTest/java/io/realm/RealmTests.java [hunk 1]
```diff
@@ -1567,6 +1572,114 @@ public void copyToRealmOrUpdate_boxedLongPrimaryKeyFieldIsNull() {
...
 
+    // @PrimaryKey + @Required annotation accept not-null value properly as a primary key value for Realm version 0.89.1+
+    @Test
+    public void copyToRealmOrUpdate_primaryKeyRequiredFieldItNotNull() {
+        final String PRIMARY_KEY_STRING_VALUE = "primaryKeyStringValue";
+        final long PRIMARY_KEY_NUMBER_VALUE = 67;
+        final String SECONDARY_FIELD_VALUE = "secondaryFieldValue";
+        final String[] PRIMARY_KEY_TYPES = {"String", "BoxedByte", "BoxedShort", "BoxedInteger", "BoxedLong"};
+
+        for (String typeName : PRIMARY_KEY_TYPES) {
```
Comment (emanuelez) [inline realm/realm-library/src/androidTest/java/io/realm/RealmTests.java]: Could this be a parametrised test instead?

Comment (stk1m1) [inline realm/realm-library/src/androidTest/java/io/realm/RealmTests.java]: Absolutely.

**Comment by stk1m1** [comment]
retest this please

**Comment by zaki50** [comment]
This PR should target `releases` branch instead of `master`.

File: realm/realm-library/src/androidTest/java/io/realm/RealmTests.java [hunk 2]
```diff
@@ -1567,6 +1572,114 @@ public void copyToRealmOrUpdate_boxedLongPrimaryKeyFieldIsNull() {
         assertEquals(SECONDARY_FIELD_UPDATED, realm.allObjects(PrimaryKeyAsBoxedLong.class).first().getName());
     }
 
+    // @PrimaryKey + @Required annotation accept not-null value properly as a primary key value for Realm version 0.89.1+
+    @Test
+    public void copyToRealmOrUpdate_primaryKeyRequiredFieldItNotNull() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmTests.java]: I would rename this to `copyToRealmOrUpdate_requiredPrimaryKey` . NotNull is implied in required, so no need to mention it twice IMO

File: realm/realm-library/src/androidTest/java/io/realm/RealmTests.java [hunk 3]
```diff
@@ -1567,6 +1572,114 @@ public void copyToRealmOrUpdate_boxedLongPrimaryKeyFieldIsNull() {
...
+                    assertEquals(SECONDARY_FIELD_VALUE, results.first().getName());
+                    break;
+                default:
+            }
+        }
+    }
+
+    // @PrimaryKey + @Required annotation does accept null as a primary key value for Realm version 0.89.1+
+    @Test
+    public void copyToRealmOrUpdate_primaryKeyRequiredFieldIsNullThrows() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmTests.java]: copyToRealmOrUpdate_requiredPrimaryKeyThrows

**Comment by cmelchior** [comment]
Missing changelog update, but apart from other comments it looks good.

**Pull request closed without merging**